ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -595,33 +595,36 @@ private void readFromFileAndPersistIfNecessary(String mountPath) throws Replicat
               FindToken token = factory.getFindToken(stream);
               // update token
               PartitionInfo partitionInfo = partitionsToReplicate.get(partitionId);
-              boolean updatedToken = false;
-              for (RemoteReplicaInfo remoteReplicaInfo : partitionInfo.getRemoteReplicaInfos()) {
-                if (remoteReplicaInfo.getReplicaId().getDataNodeId().getHostname().equalsIgnoreCase(hostname)
-                    && remoteReplicaInfo.getReplicaId().getDataNodeId().getPort() == port
-                    && remoteReplicaInfo.getReplicaId().getReplicaPath().equals(replicaPath)) {
-                  logger.info("Read token for partition {} remote host {} port {} token {}", partitionId, hostname,
-                      port, token);
-                  if (partitionInfo.getStore().getSizeInBytes() > 0) {
-                    remoteReplicaInfo.initializeTokens(token);
-                    remoteReplicaInfo.setTotalBytesReadFromLocalStore(totalBytesReadFromLocalStore);
-                  } else {
-                    // if the local replica is empty, it could have been newly created. In this case, the offset in
-                    // every peer replica which the local replica lags from should be set to 0, so that the local
-                    // replica starts fetching from the beginning of the peer. The totalBytes the peer read from the
-                    // local replica should also be set to 0. During initialization these values are already set to 0,
-                    // so we let them be.
-                    tokenWasReset = true;
-                    replicationMetrics.replicationTokenResetCount.inc();
-                    logger.info("Resetting token for partition {} remote host {} port {}, persisted token {}",
-                        partitionId, hostname, port, token);
+              if (partitionInfo != null) {
```

Comment:
That cost is unavoidable based on the current code. The token would get reset anyway if the store is offline during startup. The reset happens the first time the token is persisted (as we replace the existing token file with a newly written one, and the newly written one will not have any references about the offline store).

My suggestion was to be deterministic here and do it immediately during initialization and not leave a room where we wait till the first time the token is naturally persisted (after initialization). Just helps with the reasoning in case we need to inspect things.

If there is a scenario at all (like tests) where we would want to leave the store's token untouched, then the code that persists information has to make sure that offline store's tokens are written exactly as what was seen at startup time. I think this is an overkill at this time.